File: apps/web/src/server/auth/login-link.ts
/**
 * One-time login links — both halves, mint and consume.
 *
 * Organizer notifications deep-link into the cabinet, but `/cabinet` needs no
 * session: without one the organizer lands in the read-only *demo* cabinet
 * (ADR-010) rather than their own bookings. The messenger has already proved
 * who the recipient is, so the link carries that proof — the same argument that
 * makes `manageToken` sufficient for guests (ADR-002).
 *
 * Tokens are **minted by the notification job** at send time and consumed here
 * on `POST`. The key format, TTL and payload schema are shared through
 * `@repo/contracts` so the two halves cannot drift apart.
 *
 * Same idiom as `ticket.ts`: a peek that leaves the token alone and a
 * single-use consume. The split matters here — the landing page must be able to
 * look at a token without spending it, because link previewers and scanners
 * fetch URLs before any human does.
 */

import { randomBytes } from 'node:crypto'
import {
  LOGIN_LINK_TTL_S,
  loginLinkKey,
  type LoginLinkPayload,
  loginLinkPayload,
} from '@repo/contracts'
import { getRedis } from '@repo/redis'

import 'server-only'

/**
 * Entropy behind a login link.
 * The token is a bearer credential for a cabinet session, so it is sized to be
 * unguessable rather than short — nobody ever types it, they tap a button.
 */
const LOGIN_TOKEN_BYTES = 32

/**
 * Mint a single-use login link token pointing at `next`.
 * Called **per send attempt**, immediately before the message goes out: a retry
 * mints a fresh token and the abandoned one simply expires. That is cheaper
 * than tracking and cleaning up tokens, and it guarantees a delivered message
 * never carries a button that was already spent by an earlier attempt.
 */
export async function issueLoginLink(organizerId: string, next: string): Promise<string> {
  const token = randomBytes(LOGIN_TOKEN_BYTES).toString('base64url')
  const payload: LoginLinkPayload = { organizerId, next }

  await getRedis().set(loginLinkKey(token), JSON.stringify(payload), 'EX', LOGIN_LINK_TTL_S)

  return token
}

function parse(raw: string | null): LoginLinkPayload | null {
  if (!raw) return null

  try {
    const result = loginLinkPayload.safeParse(JSON.parse(raw))
    return result.success ? result.data : null
  } catch {
    return null
  }
}

/**
 * Read a login link **without** consuming it.
 * Used by the landing page to decide what to render. Consuming on `GET` would
 * hand the single use to Telegram's link preview crawler instead of the
 * organizer.
 */
export async function peekLoginLink(token: string): Promise<LoginLinkPayload | null> {
  return parse(await getRedis().get(loginLinkKey(token)))
}

/**
 * Atomically read **and delete** a login link.
 * `getdel` is what makes the link single-use: two concurrent submissions race
 * on one Redis command and only the winner receives a payload, so a replayed
 * POST cannot mint a second session.
 */
export async function consumeLoginLink(token: string): Promise<LoginLinkPayload | null> {
  return parse(await getRedis().getdel(loginLinkKey(token)))
}

File: apps/api-go/internal/auth/loginlink.go
package auth

import (
	"context"
	"encoding/json"
	"time"

	"api-go/internal/contracts"
	redis "api-go/internal/redis"
)

// One-time login links: notifications deep-link into the cabinet, but
// /cabinet needs no session — without one the organizer would land in
// the read-only demo cabinet (ADR-010). Minted per send attempt; a
// retry mints a fresh token and the abandoned one simply expires.

// IssueLoginLink stores {organizerId, next} behind a fresh token.
func IssueLoginLink(ctx context.Context, organizerID, next string) (string, error) {
	token := newSecretToken()
	raw, err := json.Marshal(contracts.LoginLinkPayload{OrganizerID: organizerID, Next: next})
	if err != nil {
		return "", err
	}
	if err := redis.Client().Set(ctx, contracts.LoginLinkKey(token), raw, contracts.LoginLinkTTLSeconds*time.Second).Err(); err != nil {
		return "", err
	}
	return token, nil
}

func parseLoginLink(raw string, err error) (*contracts.LoginLinkPayload, error) {
	payload, err := missingOrBroken[contracts.LoginLinkPayload](raw, err)
	if err != nil || payload == nil {
		return nil, err
	}
	// `next` is always a relative path built server-side (open-redirect
	// guard, mirrored from the loginLinkPayload schema).
	if payload.OrganizerID == "" || len(payload.Next) == 0 || payload.Next[0] != '/' {
		return nil, nil
	}
	return payload, nil
}

// PeekLoginLink reads without consuming — the landing page must be
// able to look at a token without spending it, because link previewers
// fetch URLs before any human does.
func PeekLoginLink(ctx context.Context, token string) (*contracts.LoginLinkPayload, error) {
	raw, err := redis.Client().Get(ctx, contracts.LoginLinkKey(token)).Result()
	return parseLoginLink(raw, err)
}

// ConsumeLoginLink atomically reads and deletes (GETDEL): single-use,
// so a replayed POST cannot mint a second session.
func ConsumeLoginLink(ctx context.Context, token string) (*contracts.LoginLinkPayload, error) {
	raw, err := redis.Client().GetDel(ctx, contracts.LoginLinkKey(token)).Result()
	return parseLoginLink(raw, err)
}
